File: src/test/resources/StringCompress2.java
public class Solution {
    /**
     * @param str String to compress
     * @return  Compressed string.
     */
    public String compress(String str) {

        int finalLength = countCompression(str);
        if (finalLength >= str.length()) return str;

        StringBuilder compressed = new StringBuilder(finalLength);
        int countConsecutive = 0;

        for (int i = 0; i < str.length(); i++) {
            countConsecutive++;

            if (i + 1 >= str.length() || str.charAt(i) != str.charAt(i + 1)) {
                compressed.append(str.charAt(i)).append(countConsecutive);
                countConsecutive = 0;
            }
        }

        return compressed.toString();
    }

    private static int countCompression(String str) {
        int compressedLength = 0;
        int countConsecutive = 0;
        for (int i = 0; i < str.length(); i++) {
            countConsecutive++;

            if (i + 1 >= str.length() || str.charAt(i) != str.charAt(i + 1)) {
                compressedLength += 1 + String.valueOf(countConsecutive).length();
                countConsecutive = 0;
            }
        }

        return compressedLength;
    }
}

File: src/test/resources/StringCompress2.kt
class StringCompress2 {
    /**
     * @param str String to compress
     * *
     * @return  Compressed string.
     */
    fun compress(str: String): String {

        val finalLength = countCompression(str)
        if (finalLength >= str.length) return str

        val compressed = StringBuilder(finalLength)
        var countConsecutive = 0

        for (i in 0..str.length - 1) {
            countConsecutive++

            if (i + 1 >= str.length || str[i] != str[i + 1]) {
                compressed.append(str[i]).append(countConsecutive)
                countConsecutive = 0
            }
        }

        return compressed.toString()
    }

    private fun countCompression(str: String): Int {
        var compressedLength = 0
        var countConsecutive = 0
        for (i in 0..str.length - 1) {
            countConsecutive++

            if (i + 1 >= str.length || str[i] != str[i + 1]) {
                compressedLength += 1 + countConsecutive.toString().length
                countConsecutive = 0
            }
        }

        return compressedLength
    }
}
